### ISSUE TITLE
process hangs when invoking io_uring_register_files_update in kernel 5.5.19

### ISSUE THREAD

**Issue by ShuochengWang** [axboe/liburing#256]
kernel version: 5.5.19.
When I invoke io_uring_register_files_update, the application process hangs,  its process status becomes "D, disk sleep" and I can not kill it.
In my code, 
1. the first io_uring_register_files is success
2. the first io_uring_register_files_update is success
3. the second io_uring_register_files_update **hangs**

And I find that liburing test (file-update, file-register) is success in kernel 5.5.19, while my code hangs. 
And my code is success in linux kernel 5.9.6 and 5.6.19.

My case: I have two thread (one server thread, one client thread) to use the same io_uring instance with SQPOLL. To make multithreads use the same io_uring instance, I use one lock for sq and one lock for cq.

**reproducer**:
```
#include <stdint.h>
#include <errno.h>
#include <fcntl.h>
#include <poll.h>
#include <spawn.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <netinet/in.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/wait.h>
#include <assert.h>
#include <pthread.h>

#include "liburing.h"

#define ECHO_MSG "msg for client/server test"
#define RESPONSE "ACK"
#define DEFAULT_MSG "Hello World!\n"

#define FD_NUM 16
#define RES_SIZE 10240

#define DEBUG_LOG

#define THROW_ERROR(fmt, ...)   do { \
    printf("\t\tERROR:" fmt " in func %s at line %d of file %s\n", \
    ##__VA_ARGS__, __func__, __LINE__, __FILE__); \
    return -1; \
} while (0)

struct io_uring ring;
int fds[FD_NUM];
pthread_mutex_t sq_lock;
pthread_mutex_t cq_lock;
pthread_t thread_id;
int req_id;
int has_results[RES_SIZE];
int results[RES_SIZE];

void init_io_uring() {
    int ring_flags = 0;
    // ring_flags |= IORING_SETUP_SQPOLL;
    int ret = io_uring_queue_init(1024, &ring, ring_flags);
    if (ret) {
        fprintf(stderr, "ring create failed: %d\n", ret);
    }

    for (int i = 0; i < FD_NUM; ++i) fds[i] = -1;
    ret = io_uring_register_files(&ring, fds, FD_NUM);
    if (ret) {
        fprintf(stderr, "file reg failed: %d\n", ret);
    }

    assert(pthread_mutex_init(&sq_lock, NULL) == 0);
    assert(pthread_mutex_init(&cq_lock, NULL) == 0);
    req_id = 0;
    memset(has_results, 0, sizeof(has_results));
    memset(results, 0, sizeof(results));
}

void destroy_io_uring() {
    io_uring_queue_exit(&ring);
    pthread_mutex_destroy(&sq_lock);
    pthread_mutex_destroy(&cq_lock);
}

int get_fixed_fd(int fd) {
    for (int i = 0; i < FD_NUM; ++i) {
        if (fds[i] == fd) {

            printf("fds: [");
            for (int i = 0; i < FD_NUM; ++i) printf("%d, ", fds[i]);
            printf("]\n");
            
            return i;
        }
    }
    for (int i = 0; i < FD_NUM; ++i) {
        if (fds[i] == -1) {
            printf("before io_uring_register_files_update, idx: %d, fd: %d\n", i, fd);
            int ret = io_uring_register_files_update(&ring, i, &fd, 1);
            printf("after io_uring_register_files_update\n");
            assert(ret == 1);
            fds[i] = fd;

            printf("fds: [");
            for (int i = 0; i < FD_NUM; ++i) printf("%d, ", fds[i]);
            printf("]\n");

            return i;
        }
    }
}

int submit(int fd, int flags, struct msghdr* msg, int type, const char* str) {
    pthread_mutex_lock(&sq_lock);
    #ifdef DEBUG_LOG
    printf("%s fd: %d\n", str, fd);
    #endif

    struct io_uring_sqe *sqe;

    int id = req_id++ % RES_SIZE;

    sqe = io_uring_get_sqe(&ring);
    assert(sqe != NULL);

    int fixed_fd = get_fixed_fd(fd);
    #ifdef DEBUG_LOG
    printf("%s fd: %d, fixed_fd: %d\n", str, fd, fixed_fd);
    #endif

    if (type == IORING_OP_SENDMSG)
        io_uring_prep_sendmsg(sqe, fixed_fd, msg, flags);
    else if (type == IORING_OP_RECVMSG)
        io_uring_prep_recvmsg(sqe, fixed_fd, msg, flags);
    else fprintf(stderr, "type error!");
    
    sqe->flags |= IOSQE_FIXED_FILE;
    sqe->user_data = id;

    int ret = io_uring_submit(&ring);

    #ifdef DEBUG_LOG
    printf("%s submit id (%d), ret %d, sq ready num: %d\n", str, id, ret, io_uring_sq_ready(&ring));
    #endif

    pthread_mutex_unlock(&sq_lock);

    return id;
}

int complete(int target_id, const char* str) {
    struct io_uring_cqe *cqe;
    int res = 0;
    while (1) {
        pthread_mutex_lock(&cq_lock);

        if (has_results[target_id] == 1) {
            res = results[target_id];
            pthread_mutex_unlock(&cq_lock);
            break;
        }
        else if (io_uring_peek_cqe(&ring, &cqe) == 0) {
            int cqe_data = cqe->user_data;
            int cqe_res = cqe->res;
            int cq_ready = io_uring_cq_ready(&ring);
            io_uring_cqe_seen(&ring, cqe);
            
            if (cqe_data == target_id) {
                #ifdef DEBUG_LOG
                printf("%s get self cqe, user_data %d, res %d, cq_ready: %d\n", str, cqe_data, cqe_res, cq_ready);
                #endif
                res = cqe_res;
                pthread_mutex_unlock(&cq_lock);
                break;
            }
            else {
                #ifdef DEBUG_LOG
                printf("%s get other cqe, user_data %d, res %d, want id: %d, cq_ready: %d\n", str, cqe_data, cqe_res, target_id, cq_ready);
                #endif
                has_results[cqe_data] = 1;
                results[cqe_data] = cqe_res;
            }
        }

        pthread_mutex_unlock(&cq_lock);
    }
    return res;
}

int do_read(int fd, const void* buf, size_t n) {
    struct iovec iov[1];
    iov[0].iov_base = buf;
    iov[0].iov_len = n;
    struct msghdr msg = { NULL, 0, iov, 1, NULL, 0, 0 };

    int id = submit(fd, 0, &msg, IORING_OP_RECVMSG, __func__);
    int res = complete(id, __func__);
   
    #ifdef DEBUG_LOG
    printf("%s cqe->res %d\n", __func__, res);
    #endif
    return res;
}

int do_write(int fd, const void* buf, size_t n) {
    struct iovec iov[1];
    iov[0].iov_base = buf;
    iov[0].iov_len = n;
    struct msghdr msg = { NULL, 0, iov, 1, NULL, 0, 0 };

    int id = submit(fd, 0, &msg, IORING_OP_SENDMSG, __func__);
    int res = complete(id, __func__);
    
    #ifdef DEBUG_LOG
    printf("%s cqe->res %d\n", __func__, res);
    #endif
    return res;
}

int connect_with_server(const char *addr_string, const char *port_string) {
    if (strcmp(addr_string, "NULL") == 0) {
        return 0;
    }

    int ret = 0;
    int sockfd = socket(AF_INET, SOCK_STREAM, 0);
    if (sockfd < 0) {
        THROW_ERROR("create socket error");
    }

    struct sockaddr_in servaddr;
    memset(&servaddr, 0, sizeof(servaddr));
    servaddr.sin_family = AF_INET;
    servaddr.sin_port = htons((uint16_t)strtol(port_string, NULL, 10));
    ret = inet_pton(AF_INET, addr_string, &servaddr.sin_addr);
    if (ret <= 0) {
        close(sockfd);
        THROW_ERROR("inet_pton error");
    }

    ret = connect(sockfd, (struct sockaddr *) &servaddr, sizeof(servaddr));
    if (ret < 0) {
        close(sockfd);
        THROW_ERROR("connect error");
    }

    return sockfd;
}

int client_neogotiate_msg(int server_fd, char *buf, int buf_size) {
    if (do_read(server_fd, buf, buf_size) < 0) {
        THROW_ERROR("read failed");
    }

    if (do_write(server_fd, RESPONSE, sizeof(RESPONSE)) < 0) {
        THROW_ERROR("write failed");
    }
    return 0;
}

struct client_arg {
    int port;
    char addr_string[20];
    char port_string[20];
};

int client(void* arg) {
    struct client_arg* thread_arg = arg;

    int ret = 0;
    const int buf_size = 100;
    char buf[buf_size];
    int port = thread_arg->port;
    int server_fd = connect_with_server(thread_arg->addr_string, thread_arg->port_string);

    switch (port) {
        case 8800:
            client_neogotiate_msg(server_fd, buf, buf_size);
            break;
    }

    close(server_fd);
    return ret;
}


int connect_with_child(int port, int *child_pid) {
    int ret = 0;
    int listen_fd = socket(AF_INET, SOCK_STREAM, 0);
    if (listen_fd < 0) {
        THROW_ERROR("create socket error");
    }
    int reuse = 1;
    if (setsockopt(listen_fd, SOL_SOCKET, SO_REUSEADDR, &reuse, sizeof(reuse)) < 0) {
        THROW_ERROR("setsockopt port to reuse failed");
    }

    struct sockaddr_in servaddr;
    memset(&servaddr, 0, sizeof(servaddr));
    servaddr.sin_family = AF_INET;
    servaddr.sin_addr.s_addr = htonl(INADDR_ANY);
    servaddr.sin_port = htons(port);
    ret = bind(listen_fd, (struct sockaddr *) &servaddr, sizeof(servaddr));
    if (ret < 0) {
        close(listen_fd);
        THROW_ERROR("bind socket failed");
    }

    ret = listen(listen_fd, 10);
    if (ret < 0) {
        close(listen_fd);
        THROW_ERROR("listen socket error");
    }

    char port_string[8];
    sprintf(port_string, "%d", port);
    
    struct client_arg* arg = malloc(sizeof(struct client_arg));
    arg->port = port;
    strcpy(arg->addr_string, "127.0.0.1");
    strcpy(arg->port_string, port_string);

    ret = pthread_create(&thread_id, NULL, client, (void *)arg);
    if (ret) {
        printf("unable to create thread, %d\n", ret);
    }

    int connected_fd = accept(listen_fd, (struct sockaddr *) NULL, NULL);
    if (connected_fd < 0) {
        close(listen_fd);
        THROW_ERROR("accept socket error");
    }

    close(listen_fd);
    return connected_fd;
}

int neogotiate_msg(int client_fd) {
    char buf[16];
    if (do_write(client_fd, ECHO_MSG, strlen(ECHO_MSG)) < 0) {
        THROW_ERROR("write failed");
    }

    if (do_read(client_fd, buf, sizeof(RESPONSE)) < 0) {
        THROW_ERROR("read failed");
    }

    if (strncmp(buf, RESPONSE, sizeof(RESPONSE)) != 0) {
        THROW_ERROR("msg recv mismatch");
    }
    return 0;
}

int test_read_write() {
    fprintf(stderr, "----------------------------------test_read_write---------------------------------\n");
    int ret = 0;
    int child_pid = 0;
    int client_fd = connect_with_child(8800, &child_pid);
    
    if (client_fd < 0) {
        THROW_ERROR("connect failed");
    } else {
        ret = neogotiate_msg(client_fd);
    }

    pthread_join(thread_id, NULL);

    return ret;
}


int main(int argc, const char *argv[]) {
    init_io_uring();

    test_read_write();
    
    destroy_io_uring();
}

```

the output of the reproducer:
```
do_read fd: 6
before io_uring_register_files_update, idx: 0, fd: 6
after io_uring_register_files_update
fds: [6, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, ]
do_read fd: 6, fixed_fd: 0
do_read submit id (0), ret 1, sq ready num: 0
do_write fd: 5
before io_uring_register_files_update, idx: 1, fd: 5


```

**Comment by ShuochengWang** [comment]
I use ftrace to trace the kernel function __io_uring_register, I find that it hangs in:
__io_uring_register() -> wait_for_completion() -> schedule_timeout() -> schedule() -> enter_lazy_tlb()

```
  __io_uring_register() {
    _raw_spin_lock_irqsave();
    call_rcu() {
      __call_rcu() {
        rcu_segcblist_enqueue();
      }
    }
    _raw_spin_unlock_irqrestore();
    mutex_unlock();
    wait_for_completion() {
      _cond_resched() {
        rcu_all_qs();
      }
      _raw_spin_lock_irq();
      schedule_timeout() {
        schedule() {
          rcu_note_context_switch() {
            rcu_qs();
          }
          _raw_spin_lock();
          update_rq_clock();
          deactivate_task() {
            psi_task_change() {
              record_times();
              record_times();
              record_times();
            }
            dequeue_task_fair() {
              dequeue_entity() {
                update_curr() {
                  update_min_vruntime();
                  cpuacct_charge();
                  __cgroup_account_cputime() {
                    cgroup_rstat_updated();
                  }
                }
                __update_load_avg_se();
                __update_load_avg_cfs_rq();
                clear_buddies();
                account_entity_dequeue();
                update_cfs_group();
                update_min_vruntime();
              }
              dequeue_entity() {
                update_curr() {
                  __calc_delta();
                  update_min_vruntime();
                }
                __update_load_avg_se();
                __update_load_avg_cfs_rq();
                clear_buddies();
                account_entity_dequeue();
                update_cfs_group() {
                  reweight_entity();
                }
                update_min_vruntime();
              }
              dequeue_entity() {
                update_curr() {
                  __calc_delta();
                  update_min_vruntime();
                }
                __update_load_avg_se();
                __update_load_avg_cfs_rq();
                clear_buddies();
                account_entity_dequeue();
                update_cfs_group() {
                  reweight_entity();
                }
                update_min_vruntime();
              }
              hrtick_update();
            }
          }
          pick_next_task_fair() {
            newidle_balance() {
              __msecs_to_jiffies();
            }
          }
          put_prev_task_fair() {
            put_prev_entity() {
              check_cfs_rq_runtime();
            }
            put_prev_entity() {
              check_cfs_rq_runtime();
            }
            put_prev_entity() {
              check_cfs_rq_runtime();
            }
          }
          pick_next_task_idle() {
            set_next_task_idle() {
              __update_idle_core();
            }
          }
          enter_lazy_tlb();
```

**Comment by ShuochengWang** [comment]
Is this a bug in kernel 5.5.19? since there is no problem in kernel 5.6.19 and 5.9.6.

**Comment by CarterLi** [comment]
I guess no one care about 5.5.19, since it's not the latest version, nor the LTS one

**Comment by ShuochengWang** [comment]
But 5.5.19 is the latest version of 5.5 kernel.......

**Comment by axboe** [comment]
5.5-stable was stopped 8 months ago, it's no longer being updated. Hence no fixes are going into it, it's a dead end. You should not be using dead stable kernels, they don't even get security fixes.

**Comment by axboe** [comment]
Closing this one as there's no way to get it fixed in 5.5-stable, with releases no longer being done there.
